### PR TITLE
SVCSE-1595 Setup import of tables from staging FxA databases

### PR DESCRIPTION
**Pull request by akkomar**
Per [SVCSE-1595](https://mozilla-hub.atlassian.net/browse/SVCSE-1595) we'd like to import all the tables from FxA databases to BigQuery. There are 37 tables per environment (stage and production). If we follow the current approach we'd end up having all those tables in `accounts_backend_external` dataset, and their views in `accounts_backend` together with telemetry data. This layout will make finding data hard.

To better organize this data I propose to:
* introduce two new external datasets: `accounts_db_nonprod_external` and `accounts_db_external` to replace `accounts_backend_external` and store data imported from FxA databases
* introduce two datasets to store user-facing views to these tables: `accounts_db_nonprod` and `accounts_db`
* remove `accounts_backend_external` dataset
* remove previously created user-facing views: `accounts_backend.accounts` and `accounts_backend.emails`

@kik @srose this is still work in progress, but I'd like to hear your opinions on the proposed approach. Currently this PR partially implements first step from above plan: adds `accounts_db_nonprod_external` and all the tables we'd want to import from stage.

Checklist for reviewer:

- [ ] Commits should reference a bug or github issue, if relevant (if a bug is referenced, the pull request should include the bug number in the title).
- [ ] If the PR comes from a fork, trigger integration CI tests by running the [Push to upstream workflow](https://github.com/mozilla/bigquery-etl/actions/workflows/push-to-upstream.yml) and provide the `<username>:<branch>` of the fork as parameter. The parameter will also show up
in the logs of the `manual-trigger-required-for-fork` CI task together with more detailed instructions.
- [ ] If adding a new field to a query, ensure that the schema and dependent downstream schemas have been updated.
- [ ] When adding a new derived dataset, ensure that data is not available already (fully or partially) and recommend extending an existing dataset in favor of creating new ones. Data can be available in the [bigquery-etl repository](https://github.com/mozilla/bigquery-etl), [looker-hub](https://github.com/mozilla/looker-hub) or in [looker-spoke-default](https://github.com/mozilla/looker-spoke-default/tree/e1315853507fc1ac6e78d252d53dc8df5f5f322b).

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [ ] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

[SVCSE-1595]: https://mozilla-hub.atlassian.net/browse/SVCSE-1595?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/DENG-1992)

### CODE REVIEW

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_accounts_v1/schema.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  type: STRING
+  mode: NULLABLE
+- name: email
+  type: STRING
+  mode: NULLABLE
+- name: emailVerified
+  type: BOOLEAN
+  mode: NULLABLE
+- name: verifierVersion
+  type: BOOLEAN
```
Comment (sean-rose) [inline sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_accounts_v1/schema.yaml:16]: Not all tinyints are used for boolean values.
```suggestion
- name: verifierVersion
  type: INTEGER
```

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_email_bounces_v1/schema.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  mode: NULLABLE
+- name: bounceSubType
+  type: BOOLEAN
+  mode: NULLABLE
+- name: createdAt
+  type: TIMESTAMP
+  mode: NULLABLE
+- name: emailTypeId
+  type: BOOLEAN
+  mode: NULLABLE
```
Comment (sean-rose) [inline sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_email_bounces_v1/schema.yaml:16]:
```suggestion
- name: bounceType
  type: INTEGER
  mode: NULLABLE
- name: bounceSubType
  type: INTEGER
  mode: NULLABLE
- name: createdAt
  type: TIMESTAMP
  mode: NULLABLE
- name: emailTypeId
  type: INTEGER
  mode: NULLABLE
```

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_linked_accounts_v1/schema.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+fields:
+- name: uid
+  type: STRING
+  mode: NULLABLE
+  description: Account ID in hexadecimal format.
+- name: id
+  type: STRING
+  mode: NULLABLE
+- name: providerId
+  type: BOOLEAN
```
Comment (sean-rose) [inline sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_linked_accounts_v1/schema.yaml:10]:
```suggestion
- name: providerId
  type: INTEGER
```

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_customers_v1/schema.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+fields:
+- name: uid
+  type: STRING
+  mode: NULLABLE
+  description: Account ID in hexadecimal format.
+- name: stripeCustomerId
```
Comment (sean-rose) [inline sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_customers_v1/schema.yaml:6]: I wonder if we should go to the hassle of making the BigQuery columns snake_case.  We're essentially already doing so for the table names, and if any of these tables will be made available in Looker that would impact how the column names would be displayed there (though I suppose we could also implement LookML generator logic to create snake_case dimensions for camelCase columns).

**Comment by sean-rose** [comment]
(note that because my review had more than 10 comments GitHub auto-hides some comments in the middle, which is easy to miss)

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_reset_tokens_v1/query.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+  TO_HEX(uid) AS uid,
+  SAFE.TIMESTAMP_MILLIS(SAFE_CAST(createdAt AS INT)) AS createdAt,
+FROM
+  EXTERNAL_QUERY(
+    "moz-fx-fxa-nonprod.us.fxa-rds-nonprod-stage-fxa",
+    """SELECT
+         uid,
+         createdAt
+       FROM
+         fxa.accountResetTokens
```
[the remark this answers is not in the document]
Comment (akkomar) [reply sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_reset_tokens_v1/query.sql:11]: Yeah, it's unfortunate choice of PK. Might make the table useless in BQ.

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_customers_v1/schema.yaml (hunk 4 above)
Comment (akkomar) [reply sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_customers_v1/schema.yaml:6]: I was expecting this data would be exposed in Looker via aggregates and raw tables would rather be used for exploratory analyses. One argument for keeping columns unchanged I can think of is that it makes it a bit easier to find them in FxA codebase. I'm open to discussion though.

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/metadata.yaml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  See https://mozilla.github.io/ecosystem-platform/reference/database-structure#database-fxa for more information.
+
+  Access to this dataset is restricted to accounts-confidential workgroup because some tables here contain sensitive data.
+dataset_base_acl: restricted
+user_facing: false
+labels: {}
+workgroup_access:
+- role: roles/bigquery.dataViewer
+  members:
+  - workgroup:accounts-confidential
```
[the remark this answers is not in the document]
Comment (akkomar) [reply sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/metadata.yaml:14]: Good point, but per discussion in ]the ticket](https://mozilla-hub.atlassian.net/browse/SVCSE-1595?focusedCommentId=797812) let's keep this restricted too.

**Comment by akkomar** [comment]
@sean-rose thanks for the comments. I've incorporated your suggestions in the PR, fixed some additional casting issues, and tested these queries locally.
This PR now configures import of all tables from staging environment and is ready for a final review.

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/metadata.yaml
Comment (sean-rose) [inline sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/metadata.yaml:1]: This file should be named `dataset_metadata.yaml`, not `metadata.yaml`.

**Review by sean-rose — APPROVED** [review]
r+wc

File: sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_customers_v1/schema.yaml (hunk 4 above)
Comment (sean-rose) [reply sql/moz-fx-data-shared-prod/accounts_db_nonprod_external/fxa_account_customers_v1/schema.yaml:6]: I'm fine punting on this for now.  No need to go to all that extra work if there isn't a concrete use case/benefit.

**Pull request merged**